feat(curriculum): add game character stats lab (#62243)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/blocks/lab-game-character-stats/68cabc534d2e741f33a3fe1d.md

---
id: 68cabc534d2e741f33a3fe1d
title: Build a Game Character Stats Tracker
challengeType: 27
dashedName: lab-game-character-stats
---

# --description--

In this lab, you'll build a game character stats tracker. The program will allow you to create a character with specific attributes, update those attributes, and retrieve the current stats of the character.

**Objective:** Fulfill the user stories below and get all the tests to pass to complete the lab.

**User Stories:**

1. Create a class named `GameCharacter` that represents a game character and manages character stats.

> 2. When instantiated, a new `GameCharacter` object should take a `name` parameter and initialize the following attributes:

Reviewer: It's not correct to say that an "object should take a parameter". Let's remove the mention to `name` and let campers name it whatever they want, which was why users stories have been written like this.
```suggestion
2. When instantiated, a new `GameCharacter` object should have the following attributes:
```

     - Define a setter that takes a `value` parameter, prevents health from being set below `0`, and caps the health at `100`.
5. For the `mana` property:
     - Define a getter that returns the current mana.
     - Define a setter that takes a `value` parameter, prevents mana from being set below `0`, and caps the mana at `50`.
6. Create a getter for `level` to return the character's current level.
7. Define a method named `level_up` that:
    - Increases the character's level by 1.
    - Resets health to `100` and mana to `50` using their corresponding property setters.
    - Prints a message in the form of `<name> leveled up to <level>!` (where `<name>` and `<level>` should be replaced by the character's name and new level, respectively).
8. Define a `__str__` method that returns a formatted string including:
    - The character's name.
    - The character's level.
    - The character's current health.
    - The character's current mana.
   For example, a character named `Kratos`, right after the instantiation, should be represented as the following:

   ```md
    Name: Kratos
    Level: 1
    Health: 100
    Mana: 50
   ```

## Usage example

```python

> hero = GameCharacter("Kratos") # Creates a new character named "Kratos"

Reviewer: ```suggestion
hero = GameCharacter('Kratos') # Creates a new character named Kratos
```

```js
({ 
    test: () => runPython(`assert _Node(_code).find_class("GameCharacter").find_function("health").has_args("self")`) 
})
```

You should return `self._health` in the `health` method.

```js
({ 
    test: () => runPython(`assert _Node(_code).find_class("GameCharacter").find_function("health").has_return("self._health")`)
})
```

The `health` method should have a `@property` decorator.

```js
({ 
    test: () => runPython(`assert _Node(_code).find_class("GameCharacter").find_function("health").has_decorators("property")`) 
})
```

You should create a `health` setter method.

```js
({ 
    test: () => runPython(`assert _Node(_code).find_class("GameCharacter").find_functions("health")[1]`) 
})
```

The `health` setter method should have a `@health.setter` decorator.

```js
({ 
    test: () => runPython(`assert _Node(_code).find_class("GameCharacter").find_functions("health")[1].has_decorators("health.setter")`) 
})
```

The `health` setter method should have two parameters, `self` and `value`.

```js
({ 
    test: () => runPython(`assert _Node(_code).find_class("GameCharacter").find_functions("health")[1].has_args("self, value")`) 
})
```

> You should set `self._health` to `0` if `value` is less than `0`.

Reviewer: ```suggestion
You should set `self._health` to `0` if the value given to the `health` setter is less than `0`.
```

```js
({ 
    test: () => runPython(`
    char = GameCharacter("test")
    char.health = -1
    assert char.health == 0
    char.health = -10
    assert char.health == 0
`)})
```

> You should set `self._health` to `100` if `value` is greater than `100`.

Reviewer: ```suggestion
You should set `self._health` to `100` if the value given to the `health` setter is greater than `100`.
```

    func = _Node(_code).find_class("GameCharacter").find_function("level_up")
    assert (
      func.has_call('print(f"{self._name} leveled up to {self._level}")') or
      func.has_call("print(f'{self._name} leveled up to {self._level}')")
    )
  `)
})
```

You should have a `__str__` method in the `GameCharacter` class.

```js
({ 
    test: () => runPython(`assert _Node(_code).find_class("GameCharacter").has_function("__str__")`) 
})
```

The `__str__` method should have only one parameter, `self`.

```js
({ 
    test: () => runPython(`assert _Node(_code).find_class("GameCharacter").find_function("__str__").has_args("self")`) 
})
```

You should return the character's stats with an f-string: `f"Name: {self.__name}\nLevel: {self.__level}\nHealth: {self.__health}\nMana: {self.__mana}"`.

```js
({ 
  test: () => runPython(`
    node = _Node(_code).find_class("GameCharacter").find_function("__str__")
    assert (
        node.has_return("f'Name: {self._name}\\\\nLevel: {self._level}\\\\nHealth: {self._health}\\\\nMana: {self._mana}'") or
        node.has_return('f"Name: {self._name}\\\\nLevel: {self._level}\\\\nHealth: {self._health}\\\\nMana: {self._mana}"')

>     )

Reviewer: ```suggestion
Your `__str__` method should return a string with the character's stats using the provided format.

```js
({ 
  test: () => runPython(`
import io
import sys
        
captured_output = io.StringIO()
sys.stdout = captured_output
        
gc = GameCharacter("test")
gc.mana = 15
gc.health = 77
print(gc)
        
sys.stdout = sys.__stdout__
output = captured_output.getvalue()

expected = """Name: test
Level: 1
Health: 77
Mana: 15"""
assert expected in output
```
